pkg/promtail/client/client.go

```diff
@@ -123,49 +131,84 @@ func (c *Client) run() {
 				batch[fp] = stream
 			}
 			stream.Entries = append(stream.Entries, e.Entry)
+
 		case <-maxWait.C:
 			if len(batch) > 0 {
-				if err := c.send(batch); err != nil {
-					level.Error(c.logger).Log("msg", "error sending batch", "error", err)
-				}
+				c.sendBatch(batch)
 				batchSize = 0
 				batch = map[model.Fingerprint]*logproto.Stream{}
 			}
 		}
 	}
 }
 
-func (c *Client) send(batch map[model.Fingerprint]*logproto.Stream) error {
+func (c *Client) sendBatch(batch map[model.Fingerprint]*logproto.Stream) {
+	buf, err := encodeBatch(batch)
+	if err != nil {
+		level.Error(c.logger).Log("msg", "error encoding batch", "error", err)
+		return
+	}
+	sentBytes.Add(float64(len(buf)))
+
+	ctx := context.Background()
+	backoff := util.NewBackoff(ctx, c.cfg.BackoffConfig)
+	for backoff.Ongoing() {
+		start := time.Now()
+		status, err := c.send(ctx, buf)
+		requestDuration.WithLabelValues(strconv.Itoa(status)).Observe(time.Since(start).Seconds())
+		if err == nil {
+			return
+		}
+
+		level.Error(c.logger).Log("msg", "error sending batch", "status", status, "error", err)
+
+		// Only retry 500s and connection-level errors.
+		if status > 0 && status/100 != 5 {
+			return
+		}
+
```

Comment:
What about logging a Warn if the status is 5xx and we are still retrying, then log an error when we run out of retries, and log any non-retryable error immediately as Error?

This way we when we see Error in the logs we know we are dropping messages, however, Warns would give us some indicator of transient issues.